File: Vantage/Services/Providers/ConnectionTest.swift
import Foundation

/// Verifies a stored credential by making the cheapest real request each
/// provider offers.
///
/// A key can only be confirmed by using it. Storing one and discovering hours
/// later that a chart is empty — because a provider answered 403 for a mistyped
/// token — is exactly the failure this removes.
enum ConnectionTest {
    enum Result: Sendable, Equatable {
        case success(detail: String)
        case failure(APIError)

        var isSuccess: Bool { if case .success = self { return true }; return false }

        var message: String {
            switch self {
            case .success(let detail): detail
            case .failure(let error):
                error.recoverySuggestion.map { "\(error.shortDescription) — \($0)" }
                    ?? error.shortDescription
            }
        }
    }

    static func run(
        for provider: DataProviderID,
        registry: ProviderRegistry
    ) async -> Result {
        do {
            switch provider {
            case .finnhub:
                let quote = try await registry.marketData.quote(symbol: "AAPL")
                return .success(detail: "Live quote received (AAPL \(Format.currency(quote.last))).")

            case .tiingo:
                let bars = try await registry.marketData.bars(
                    symbol: "AAPL", resolution: .daily,
                    from: Date.now.addingTimeInterval(-10 * 86_400), to: .now
                )
                guard !bars.isEmpty else {
                    return .failure(.noData(.tiingo, endpoint: "tiingo prices"))
                }
                return .success(detail: "\(bars.count) daily bars received.")

            case .alpaca:
                // Asked for over the last four days so the window spans a
                // weekend without the test reading as a failure on a Monday.
                let intraday = try await registry.marketData.bars(
                    symbol: "AAPL", resolution: .fifteenMinute,
                    from: Date.now.addingTimeInterval(-4 * 86_400), to: .now
                )
                guard !intraday.isEmpty else {
                    return .failure(.noData(.alpaca, endpoint: "alpaca bars"))
                }
                return .success(detail: "\(intraday.count) 15-minute IEX bars received.")

            case .fred:
                guard let macro = registry.macro else {
                    return .failure(.missingCredentials(.fred))
                }
                let observations = try await macro.observations(
                    seriesID: "SP500",
                    from: Date.now.addingTimeInterval(-10 * 86_400), to: .now
                )
                guard let latest = observations.last else {
                    return .failure(.noData(.fred, endpoint: "SP500"))
                }
                return .success(detail: "S&P 500 at \(Format.ratio(latest.value, precision: 2)).")

            case .sec:
                guard let sec = registry.sec else {
                    return .failure(.missingCredentials(.sec))
                }
                // Resolving a well-known ticker exercises the whole path: the
                // User-Agent EDGAR demands, the ticker map, and decoding.
                let cik = try await sec.resolveCIK(symbol: "AAPL")
                let filings = try await sec.filings(cik: cik, formTypes: [], limit: 5)
                return .success(
                    detail: "CIK \(cik) resolved, \(filings.count) recent filings."
                )

            case .computed, .sample:
                // Neither is a connection. `.sample` reaches nothing, and a
                // test that "succeeded" against invented data would report the
                // opposite of what it measured.
                return .failure(.noData(provider, endpoint: "connection test"))
            }
        } catch let error as APIError {
            return .failure(error)
        } catch {
            return .failure(.transport(provider, underlying: error.localizedDescription))
        }
    }
}

File: kmp/core/src/commonMain/kotlin/com/tylerabitbol/libra/services/providers/ConnectionTest.kt
package com.tylerabitbol.libra.services.providers

import com.tylerabitbol.libra.models.core.BarResolution
import com.tylerabitbol.libra.models.provenance.DataProviderID
import com.tylerabitbol.libra.networking.APIError
import com.tylerabitbol.libra.support.Format
import kotlinx.coroutines.CancellationException
import kotlin.time.Clock
import kotlin.time.Duration.Companion.days

/**
 * Verifies a stored credential by making the cheapest real request each
 * provider offers.
 *
 * A key can only be confirmed by using it. Storing one and discovering hours
 * later that a chart is empty — because a provider answered 403 for a mistyped
 * token — is exactly the failure this removes.
 */
object ConnectionTest {

    sealed class Result {
        data class Success(val detail: String) : Result()
        data class Failure(val error: APIError) : Result()

        val isSuccess: Boolean get() = this is Success

        val message: String
            get() = when (this) {
                is Success -> detail
                is Failure -> error.recoverySuggestion
                    ?.let { "${error.shortDescription} — $it" }
                    ?: error.shortDescription
            }
    }

    suspend fun run(provider: DataProviderID, registry: ProviderRegistry): Result = try {
        val now = Clock.System.now()
        when (provider) {
            DataProviderID.Finnhub -> {
                val quote = registry.marketData.quote("AAPL")
                Result.Success("Live quote received (AAPL ${Format.currency(quote.last)}).")
            }

            DataProviderID.Tiingo -> {
                val bars = registry.marketData.bars(
                    "AAPL", BarResolution.Daily, now - 10.days, now,
                )
                if (bars.isEmpty()) {
                    Result.Failure(
                        APIError.NoData(DataProviderID.Tiingo, endpoint = "tiingo prices"),
                    )
                } else {
                    Result.Success("${bars.size} daily bars received.")
                }
            }

            DataProviderID.Alpaca -> {
                // Asked for over the last four days so the window spans a
                // weekend without the test reading as a failure on a Monday.
                val intraday = registry.marketData.bars(
                    "AAPL", BarResolution.FifteenMinute, now - 4.days, now,
                )
                if (intraday.isEmpty()) {
                    Result.Failure(
                        APIError.NoData(DataProviderID.Alpaca, endpoint = "alpaca bars"),
                    )
                } else {
                    Result.Success("${intraday.size} 15-minute IEX bars received.")
                }
            }

            DataProviderID.FRED -> {
                val macro = registry.macro
                    ?: return Result.Failure(APIError.MissingCredentials(DataProviderID.FRED))
                val observations = macro.observations("SP500", now - 10.days, now)
                val latest = observations.lastOrNull()
                    ?: return Result.Failure(
                        APIError.NoData(DataProviderID.FRED, endpoint = "SP500"),
                    )
                Result.Success("S&P 500 at ${Format.ratio(latest.value, precision = 2)}.")
            }

            DataProviderID.SEC -> {
                val sec = registry.sec
                    ?: return Result.Failure(APIError.MissingCredentials(DataProviderID.SEC))
                // Resolving a well-known ticker exercises the whole path: the
                // User-Agent EDGAR demands, the ticker map, and decoding.
                val cik = sec.resolveCIK("AAPL")
                val filings = sec.filings(cik, formTypes = emptyList(), limit = 5)
                Result.Success("CIK $cik resolved, ${filings.size} recent filings.")
            }

            DataProviderID.Computed, DataProviderID.Sample ->
                // Neither is a connection. `Sample` reaches nothing, and a test
                // that "succeeded" against invented data would report the
                // opposite of what it measured.
                Result.Failure(APIError.NoData(provider, endpoint = "connection test"))
        }
    } catch (cancellation: CancellationException) {
        throw cancellation
    } catch (error: APIError) {
        Result.Failure(error)
    } catch (error: Exception) {
        Result.Failure(
            APIError.Transport(provider, underlying = error.message ?: "unknown error"),
        )
    }
}
